src/common/pf_alarm.c

```diff
@@ -2267,74 +2236,296 @@ static int pf_alarm_send_internal (
    return ret;
 }
 
+/************************ Queue handling ************************************/
+
 /**
- * Reset queues for outgoing alarms
- * @param net              InOut: The p-net stack instance
- * @param p_ar             InOut: The AR instance.
- * @return 0 is always returned
+ * @internal
+ * Create a mutex for a queue.
+ *
+ * Note that on first usage, the pf_queue_accountant_t must be fully cleared.
+ *
+ * @param p_accountant     InOut: Queue accountant
  */
-static int pf_alarm_reset_send_queues (pnet_t * net, pf_ar_t * p_ar)
+void pf_alarm_queue_mutex_create (pf_queue_accountant_t * p_accountant)
 {
-   memset (p_ar->alarm_send_q, 0, sizeof (p_ar->alarm_send_q));
+   if (p_accountant->mutex == NULL)
+   {
+      p_accountant->mutex = os_mutex_create();
+   }
+}
+
+/**
+ * @internal
+ * Destroy a mutex for a queue.
+ *
+ * @param p_accountant     InOut: Queue accountant
+ */
+void pf_alarm_queue_mutex_destroy (pf_queue_accountant_t * p_accountant)
+{
+   if (p_accountant->mutex != NULL)
+   {
+      os_mutex_destroy (p_accountant->mutex);
+      p_accountant->mutex = NULL;
+   }
+}
+
+/**
+ * @internal
+ * Reset the queue indexes.
+ *
+ * Locks the queue.
+ *
+ * @param p_accountant     InOut: Queue accountant
+ */
+static void pf_alarm_queue_lock_and_reset (pf_queue_accountant_t * p_accountant)
+{
+   assert (p_accountant->mutex != NULL);
+   os_mutex_lock (p_accountant->mutex);
+   p_accountant->count = 0;
+   p_accountant->read_index = 0;
+   p_accountant->write_index = 0;
+}
+
+/**
+ * @internal
+ * Unlock a queue.
+ *
+ * @param p_accountant     InOut: Queue accountant
+ */
+static void pf_alarm_queue_unlock (pf_queue_accountant_t * p_accountant)
+{
+   assert (p_accountant->mutex != NULL);
+   os_mutex_unlock (p_accountant->mutex);
+}
+
+/**
+ * @internal
+ * Get next write index for a queue.
+ *
+ * Locks the queue on success.
+ *
+ * @param p_accountant     InOut: Queue accountant
+ * @param p_write_index    Out:   Write index to use
+ * @return 0 if there is room in the queue (Queue will be locked)
+ *         -1 if the queue is full (Will not be locked)
+ */
+static int pf_alarm_queue_lock_and_get_writeindex (
+   pf_queue_accountant_t * p_accountant,
+   uint16_t * p_write_index)
+{
+   os_mutex_lock (p_accountant->mutex);
+   if (p_accountant->count >= PNET_MAX_ALARMS)
+   {
+      os_mutex_unlock (p_accountant->mutex);
```

Comment:
This function does not always exit with the mutex locked. That leads to confusing code later on, where it looks like calling functions forgot to unlock the mutex. Could the code be organised in a way that avoids the "lock_and"? Once you have allocated the write-index, do you need to keep the lock?

Reply:
Restructured the logic.
Done